docs/internal/expressions.md

```diff
@@ -0,0 +1,611 @@
+
+Verona does not have _free functions_, ie. those that are not attached to a type.
+In that sense, all Verona functions are methods of some class, for example, like `U32::+(U32, U32)` implements addition of unsigned 32-bit integers for `(a + b)`.
+
+Modules are classes, so they can have functions without declaring further types, but those are just methods of the module's class type.
+So a file can have functions outside of declared types that look like free functions, but to call them, one must either pass the full name (ex. `MyModule::function()`) or `use` a module to import that module's context on your own.
+
+Example:
+```ts
+// On MyModule
+{
+  ...
+  foo() { ... }
+}
+
+// On OtherModule
+{
+  ...
+  foo() { ... }
+}
+
+// On some other module
+using "MyModule";
+bar()
+{
+  foo(); // This is fine, since only MyModule's scope has been imported here
+}
+
+// On some yet another module
+using "MyModule";
+using "OtherModule";
+bar()
+{
+  foo(); // ERROR: which one?
+  MyModule::foo(); // Fully qualified, this is fine.
+}
+
+// On the final module, I promise
+using "MyModule";
+type Other = "OtherModule";
+bar()
+{
+  foo(); // Fine, `OtherModule`'s version is `Other::foo()`.
+  Other::foo(); // Fully qualified, using type alias, this is fine.
+}
+```
+
+### Calling methods
+
+Calling a method by name must allow the compiler to find the class or interface it belongs to at compile time.
+For concrete class types, the call will be static (direct call to the function), and for abstract interface or union types, a dynamic dispatch will be performed to the object's virtual table.
+
+If the method has return values, the variable (named or temporary) that will hold the results must be of a compatible type, declared or inferred, with the rest of the code that uses it.
+
+Example:
+```ts
+my_function()
+{
+  // An object of type `MyType` (see above)
+  let a : MyType & mut = MyType;
+
+  // You can call methods via an object
+  // Here, the first argument becomes `a`
+  // Because we know the concrete type, this is a static call
+  let b = a.my_method();
+
+  // Or statically, with an explicit first argument
+  // This is identical to the one above
+  let c = MyType::my_method(a);
+
+  // Or use infix notation, equivalent to `MyType::===(U32, U32)`
+  let d : MyType & mut = MyType;
+  let (equals, result, error) = (d === 42);
+  // `equals` is a `Bool`, `result` is `MyType` and `error` is a `MaybeError`
+}
+```
+
+The notations used for calling functions are interchangeable and do not force semantics change (dynamic vs. static) calling, which will only be determined by being able to determine, at compile time, if the type is concrete or not.
+
+_Note: It's still under review if we'll use named parameters or forwarding arguments for calls._
+
+_Note2: Check the type document for a review on static vs. dynamic methods, constructors and destructors._
+
+## Equality
+
+Verona has two types of equality: by address and structural.
+
+### Address equality
```

Comment:
Given this week's discussion, I think this section probably needs some significant rewrites.

Reply:
Yes, I'd like to get this right before merge, tbh.